src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1603,112 +1646,98 @@ public RefreshItem(String keyInCache, AuthenticationRequest request, TokenCacheI
-            if (isFamilyClientIdFeatureEnabled) {
-                // If still cannot find an item, try with family clientid
-                // Disable this feature for ADFS authority
-                if (item == null || StringExtensions.IsNullOrBlank(item.getRefreshToken())) {
-                    if (!UrlExtensions.isADFSAuthority(StringExtensions.getUrl(mAuthority))) {
-                        Logger.v(TAG + methodName, "No refresh token found, trying to find family client id item.");
-                        item = findFamilyItemForUser(request.getUserIdentifierType(), userId);
-                    } else {
-                        Logger.v(TAG + methodName, "No refresh token found, skip the family client id item lookup for ADFS authority.");
+            Logger.v(TAG + methodName, "Look for resource specific refresh token in the cache.");
+            refreshItem = findResourceSpecificRefreshToken(request);
+            
+            // If regular RT is found, but it's not MRRT, shouldn't bother trying with FRT. 
+            // All the token issued by EvoSts is MRRT. If regular RT is not MRRT, it basically means
+            // it's not issued by EvoSts. FoCI feature is only visible to RT issued by evoSTS. 
+            if (refreshItem == null || refreshItem.mMultiResource == true) {
+                Logger.v(TAG + methodName, "Resource specific RT does not exist or it is also a MRRT, "
+                        + "look for MRRT.");
+                refreshItem = findMultiResourceRefreshToken(request);
+                
+                if (refreshItem == null || refreshItem.mFamilyToken == true) {
+                    Logger.v(TAG + methodName, "MRRT does not exist or MRRT is also a FRT, "
+                            + "look for FRT.");
+                    refreshItem = findFamilyRefreshToken(request);
+                    
+                    if (refreshItem == null) {
+                        Logger.v(TAG + methodName, "FRT does not exist, fall back to MRRT if exists.");
+                        refreshItem = findMultiResourceRefreshToken(request);
                     }
                 }
             }
-
-            if (item != null && !StringExtensions.IsNullOrBlank(item.getRefreshToken())) {
-                String refreshTokenHash = getTokenHash(item.getRefreshToken());
-
-                Logger.v(TAG, "Refresh token is available and id:" + refreshTokenHash
-                        + " Key used:" + keyUsed);
-                refreshItem = new RefreshItem(keyUsed, request, item);
-            }
         }
 
         return refreshItem;
     }
-
-    private TokenCacheItem findFamilyItemForUser(final UserIdentifierType userIdentifierType, final String userId) {
-        final String methodName = ":findFamilyItemForUser";
-            
+    
+    private RefreshItem findResourceSpecificRefreshToken(final AuthenticationRequest authRequest) {
+        final String userId = getUserFromAuthenticationRequest(authRequest);
```

Comment:
Method getRefreshTokenFromCache must have only 2 arguments:
userId (which already detected as userId or dispayable)
and enum value which key you're like to generate
do not create separate method for each refresh token